chain-impl-mockchain/src/vote/manager.rs

```diff
@@ -434,24 +455,34 @@ impl ProposalManagers {
 
     /// validate the vote against the proposal: verify that the proposal exists
     /// and the the length of the ciphertext is correct (if applicable)
-    pub fn validate_vote(&self, cast: &VoteCast) -> Result<(), VoteError> {
+    pub fn validate_vote(
+        &self,
+        cast: VoteCast,
+        vote_plan: &VotePlan,
+    ) -> Result<(ValidatedVote, usize), VoteError> {
         let proposal_index = cast.proposal_index() as usize;
         if let Some(manager) = self.0.get(proposal_index) {
-            manager.validate_vote(cast)
+            Ok((manager.validate_vote(cast, vote_plan)?, proposal_index))
         } else {
             Err(VoteError::InvalidVoteProposal {
                 num_proposals: self.0.len(),
-                vote: cast.clone(),
+                vote: cast,
             })
         }
     }
 
-    pub fn start_private_tally(&self, stake: &StakeControl) -> Result<Self, VoteError> {
+    pub fn start_private_tally(
+        &self,
+        stake: &StakeControl,
+        vote_plan: &VotePlan,
```

Comment:
Why pass the vote plan in a method to a `ProposalManagers` instance that's already associated with supposedly the same vote plan by construction? I think it's better to embed the CRS and public key data into the object on construction.
And maybe split the proposal managers into public and private kinds so that you don't end up with a lot of `Option` to accommodate public vote plans in the same struct.